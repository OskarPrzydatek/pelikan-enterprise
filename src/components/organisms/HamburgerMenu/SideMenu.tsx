import React from 'react';

import { Background, Button, Header, Icon } from '~/components/atoms';
import { Icons, Slugs } from '~/constants';
import { sideMenuNavigationItemsMetadata } from '~/metadata';
import { IComponent, ISideMenuNavigationList } from '~/models';

import * as S from './SideMenu.styles';

interface ISideMenuNavigationItem extends ISideMenuNavigationList {
  handleIsCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
}

const SideMenuNavigationItem: React.FC<ISideMenuNavigationItem> = ({
  title,
  items,
  handleIsCurrentPage,
  onNavigate,
}: ISideMenuNavigationItem) => {
  const handleIsCurrentPageStyle = (path: Slugs) =>
    handleIsCurrentPage(path) ? S.currentPageButtonCSS : S.pageButtonCSS;

  return (
    <S.NavItem>
      <Header>{title}</Header>
      {items.map(({ path, label }) => (
        <Button
          key={path}
          css={handleIsCurrentPageStyle(path)}
          dataTestID={`side-menu-item-${path}`}
          variant="ghost"
          onClick={() => onNavigate(path)}
        >
          {label}
        </Button>
      ))}
    </S.NavItem>
  );
};

interface ISideMenu extends IComponent {
  onClickCloseSideMenu: () => void;
  handleIsCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
  onClickLogout: () => void;
}

export const SideMenu: React.FC<ISideMenu> = ({
  onClickCloseSideMenu,
  handleIsCurrentPage,
  onNavigate,
  onClickLogout,
}: ISideMenu) => {
  return (
    <Background onClickBackground={onClickCloseSideMenu}>
      <S.SideMenu data-testid="hamburger-menu-side-menu">
        <S.HeaderBar>
          <Header>Pelikan Enterprise</Header>
          <Icon
            dataTestID="side-menu-close-icon"
            icon={Icons.CLOSE}
            width={15}
            onClick={onClickCloseSideMenu}
          />
        </S.HeaderBar>
        <S.Nav>
          <S.List>
            {sideMenuNavigationItemsMetadata.map(({ title, items }) => (
              <SideMenuNavigationItem
                key={title}
                handleIsCurrentPage={handleIsCurrentPage}
                items={items}
                title={title}
                onNavigate={onNavigate}
              />
            ))}
          </S.List>
        </S.Nav>
        <Button
          css={S.logoutButtonCSS}
          dataTestID="side-menu-logout-button"
          onClick={onClickLogout}
        >
          Wyloguj
        </Button>
      </S.SideMenu>
    </Background>
  );
};
