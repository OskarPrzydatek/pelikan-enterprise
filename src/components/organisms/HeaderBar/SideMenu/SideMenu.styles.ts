import styled, { css } from 'styled-components';

export const SideMenu = styled.div`
  height: 100vh;
  width: 100%;
  background-color: ${({ theme }) => theme.colors.white};
  display: grid;
  grid-template-rows: 1fr 10fr 0.5fr;
  padding: ${({ theme }) => theme.spaces.s};

  @media ${({ theme }) => theme.breakpoints.phone} {
    width: 55%;
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    width: 30%;
  }

  @media ${({ theme }) => theme.breakpoints.desktop} {
    width: 20%;
  }
`;

export const HeaderBar = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const Nav = styled.nav``;

export const List = styled.ul``;

export const NavItem = styled.li`
  display: flex;
  flex-direction: column;

  button {
    margin-left: ${({ theme }) => theme.spaces.xl};
    font-weight: 700;
    cursor: pointer;
  }
`;

export const appNameLabelCSS = css`
  color: ${({ theme }) => theme.colors.darkBlue};
  font-weight: 900;
`;

export const currentPageButtonCSS = css`
  color: ${({ theme }) => theme.colors.lightBlue};
`;

export const pageButtonCSS = css`
  color: ${({ theme }) => theme.colors.darkBlue};
`;

export const logoutButtonCSS = css`
  color: ${({ theme }) => theme.colors.white};
`;
