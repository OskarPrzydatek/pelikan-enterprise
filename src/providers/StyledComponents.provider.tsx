import { ThemeProvider } from 'styled-components';

import { GlobalStyleLayout } from '../styles/layouts';
import { theme } from '../styles/theme';

interface IStyledComponentsProvider {
  children: React.ReactNode;
}

export const StyledComponentsProvider: React.FC<IStyledComponentsProvider> = ({
  children,
}: IStyledComponentsProvider) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyleLayout />
      {children}
    </ThemeProvider>
  );
};
