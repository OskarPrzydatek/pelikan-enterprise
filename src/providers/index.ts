export { StyledComponentsProvider } from './StyledComponents.provider';
