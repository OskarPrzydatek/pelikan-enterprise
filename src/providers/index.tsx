import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { GlobalStyleLayout } from '../styles/layouts';
import { theme } from '../styles/theme';

interface IStyledComponentsProvider {
  children: React.ReactNode;
}

interface ITestingPageProvider {
  element: JSX.Element;
}

export const StyledComponentsProvider: React.FC<IStyledComponentsProvider> = ({
  children,
}: IStyledComponentsProvider) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyleLayout />
      {children}
    </ThemeProvider>
  );
};

export const TestingPageProvider: React.FC<ITestingPageProvider> = ({
  element,
}: ITestingPageProvider) => {
  return (
    <StyledComponentsProvider>
      <MemoryRouter initialEntries={['/page1']}>
        <Routes>
          <Route element={element} path="/:page" />
        </Routes>
      </MemoryRouter>
    </StyledComponentsProvider>
  );
};
