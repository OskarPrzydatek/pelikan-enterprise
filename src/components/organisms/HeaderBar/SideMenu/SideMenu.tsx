import React from 'react';

import { Background, Button, Header, Icon, Text } from '~/components/atoms';
import { Icons, Slugs } from '~/constants';
import { IComponent } from '~/models';
import { ILoggedUser } from '~/models/data';

import * as S from './SideMenu.styles';

interface ISideMenuItem {
  slug: Slugs;
  label: string;
  handleIsCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
}

const SideMenuItem: React.FC<ISideMenuItem> = ({
  slug,
  label,
  handleIsCurrentPage,
  onNavigate,
}: ISideMenuItem) => {
  const handleNavigate = () => onNavigate(slug);
  const handleCurrentPageCSS = (slug: Slugs) =>
    handleIsCurrentPage(slug) ? S.currentPageButtonCSS : S.pageButtonCSS;

  return (
    <S.NavItem>
      <Button
        dataTestID={`side-menu-item-${slug}`}
        variant="ghost"
        onClick={handleNavigate}
      >
        <Header css={handleCurrentPageCSS(slug)}>{label}</Header>
      </Button>
    </S.NavItem>
  );
};

interface ISideMenu extends IComponent {
  onClickCloseSideMenu: () => void;
  handleIsCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
  onClickLogout: () => void;
}

export const SideMenu: React.FC<ISideMenu> = ({
  onClickCloseSideMenu,
  handleIsCurrentPage,
  onNavigate,
  onClickLogout,
}: ISideMenu) => {
  const loggedUser = JSON.parse(
    localStorage.getItem('loggedUser')!
  ) as ILoggedUser;

  return (
    <Background onClickBackground={onClickCloseSideMenu}>
      <S.SideMenu data-testid="hamburger-menu-side-menu">
        <S.HeaderBar>
          <Text css={S.appNameLabelCSS}>Pelikan Enterprise</Text>
          <Icon
            dataTestID="side-menu-close-icon"
            icon={Icons.CLOSE}
            width={15}
            onClick={onClickCloseSideMenu}
          />
        </S.HeaderBar>
        <S.Nav>
          <S.List>
            <SideMenuItem
              handleIsCurrentPage={handleIsCurrentPage}
              label="Oferty"
              slug={Slugs.OFFER_OVERVIEW}
              onNavigate={onNavigate}
            />
            <SideMenuItem
              handleIsCurrentPage={handleIsCurrentPage}
              label="Hotele"
              slug={Slugs.HOTEL_OVERVIEW}
              onNavigate={onNavigate}
            />
            <SideMenuItem
              handleIsCurrentPage={handleIsCurrentPage}
              label="Transport"
              slug={Slugs.TRANSPORT_OVERVIEW}
              onNavigate={onNavigate}
            />
            <SideMenuItem
              handleIsCurrentPage={handleIsCurrentPage}
              label="Atrakcje"
              slug={Slugs.ATTRACTION_OVERVIEW}
              onNavigate={onNavigate}
            />
            <SideMenuItem
              handleIsCurrentPage={handleIsCurrentPage}
              label="Hashtagi"
              slug={Slugs.HASHTAG_OVERVIEW}
              onNavigate={onNavigate}
            />
            {loggedUser.userType === 'ADMIN' ? (
              <SideMenuItem
                handleIsCurrentPage={handleIsCurrentPage}
                label="UÅ¼ytkownicy"
                slug={Slugs.USERS_OVERVIEW}
                onNavigate={onNavigate}
              />
            ) : null}
          </S.List>
        </S.Nav>
        <Button
          css={S.logoutButtonCSS}
          dataTestID="side-menu-logout-button"
          onClick={onClickLogout}
        >
          Wyloguj
        </Button>
      </S.SideMenu>
    </Background>
  );
};
