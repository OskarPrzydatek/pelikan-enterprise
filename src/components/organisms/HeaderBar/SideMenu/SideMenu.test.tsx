import { fireEvent, render, screen } from '@testing-library/react';
import { describe, test, expect, vi } from 'vitest';

import { StyledComponentsProvider } from '~/providers';

import { SideMenu } from './SideMenu';

const mockHandleIsCurrentPage = vi.fn();
const mockOnClickCloseSideMenu = vi.fn();
const mockOnClickLogout = vi.fn();
const mockOnNavigate = vi.fn();

const MockSideMenu = () => (
  <StyledComponentsProvider>
    <SideMenu
      handleIsCurrentPage={mockHandleIsCurrentPage}
      onClickCloseSideMenu={mockOnClickCloseSideMenu}
      onClickLogout={mockOnClickLogout}
      onNavigate={mockOnNavigate}
    />
  </StyledComponentsProvider>
);

const loggedUserMock = {
  firstName: 'fname',
  lastName: 'lname',
  userType: 'WORKER',
};

beforeEach(() => {
  localStorage.setItem('loggedUser', JSON.stringify(loggedUserMock));
});

describe('SideMenu', () => {
  test('component snapshot', () => {
    const view = render(<MockSideMenu />);
    expect(view).toMatchSnapshot();
  });

  test('ensure navigation works correctly', () => {
    render(<MockSideMenu />);
    fireEvent.click(screen.getByTestId('side-menu-item-offer-overview'));
    expect(mockOnNavigate).toHaveBeenCalledWith('offer-overview');
  });

  test('ensure logout works correctly', () => {
    render(<MockSideMenu />);
    fireEvent.click(screen.getByTestId('side-menu-logout-button'));
    expect(mockOnClickLogout).toHaveBeenCalled();
  });
});
