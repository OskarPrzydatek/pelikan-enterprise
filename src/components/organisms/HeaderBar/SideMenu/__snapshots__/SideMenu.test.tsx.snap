// Vitest Snapshot v1

exports[`SideMenu > component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div
      class="sc-dkrFOg hVytYR"
    >
      <div
        class="sc-jrcTuL kapCbJ"
        data-testid="hamburger-menu-side-menu"
      >
        <div
          class="sc-kDvujY bjUeDl"
        >
          <p
            class="sc-pyfCe OkzBx"
          >
            Pelikan Enterprise
          </p>
          <button
            class="sc-jSUZER eNWyGN"
            data-testid="side-menu-close-icon"
          >
            <svg
              data-testid="close-icon"
              fill="none"
              viewBox="0 0 20 20"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 2.01429L17.9857 0L10 7.98571L2.01429 0L0 2.01429L7.98571 10L0 17.9857L2.01429 20L10 12.0143L17.9857 20L20 17.9857L12.0143 10L20 2.01429Z"
                fill="#013059"
              />
            </svg>
          </button>
        </div>
        <nav
          class="sc-ipEyDJ"
        >
          <ul
            class="sc-csuSiG"
          >
            <li
              class="sc-eDWCr xPKqo"
            >
              <button
                class="sc-hLBbgP grCgBk"
                data-testid="side-menu-item-offer-overview"
              >
                <h2
                  class="sc-eDvSVe jPXdpL"
                >
                  Oferty
                </h2>
              </button>
            </li>
            <li
              class="sc-eDWCr xPKqo"
            >
              <button
                class="sc-hLBbgP grCgBk"
                data-testid="side-menu-item-hotel-overview"
              >
                <h2
                  class="sc-eDvSVe jPXdpL"
                >
                  Hotele
                </h2>
              </button>
            </li>
            <li
              class="sc-eDWCr xPKqo"
            >
              <button
                class="sc-hLBbgP grCgBk"
                data-testid="side-menu-item-transport-overview"
              >
                <h2
                  class="sc-eDvSVe jPXdpL"
                >
                  Transport
                </h2>
              </button>
            </li>
            <li
              class="sc-eDWCr xPKqo"
            >
              <button
                class="sc-hLBbgP grCgBk"
                data-testid="side-menu-item-attraction-overview"
              >
                <h2
                  class="sc-eDvSVe jPXdpL"
                >
                  Atrakcje
                </h2>
              </button>
            </li>
            <li
              class="sc-eDWCr xPKqo"
            >
              <button
                class="sc-hLBbgP grCgBk"
                data-testid="side-menu-item-hashtag-overview"
              >
                <h2
                  class="sc-eDvSVe jPXdpL"
                >
                  Hashtagi
                </h2>
              </button>
            </li>
          </ul>
        </nav>
        <button
          class="sc-hLBbgP dFUGmA"
          data-testid="side-menu-logout-button"
        >
          Wyloguj
        </button>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
